httpclient/httpclient.go

```diff
@@ -26,3 +31,113 @@ func CloneHTTPHeader(in http.Header) http.Header {
 	}
 	return out
 }
+
+// ClientProviders for further customization of the client logging and request id.
+type ClientProviders struct {
+	// Logger is a function that provides a context-specific logger.
+	Logger func(ctx context.Context) log.FieldLogger
+
+	// RequestID is a function that provides a request ID.
+	RequestID func(ctx context.Context) string
+}
+
+// NewHTTPClient wraps delegate transports with logging, metrics, rate limiting, retryable, user agent, request id
+// and returns an error if any occurs.
+func NewHTTPClient(
+	cfg *Config,
+	userAgent string,
+	reqType string,
+	delegate http.RoundTripper,
+	providers ClientProviders,
+) (*http.Client, error) {
+	var err error
+
+	if delegate == nil {
+		delegate = http.DefaultTransport.(*http.Transport).Clone()
+	}
+
+	if cfg.Logger.Enabled {
+		opts := cfg.Logger.TransportOpts()
+		opts.LoggerProvider = providers.Logger
+		delegate = NewLoggingRoundTripperWithOpts(delegate, reqType, opts)
+	}
+
+	if cfg.Metrics.Enabled {
+		delegate = NewMetricsRoundTripper(delegate, reqType)
+	}
+
+	if cfg.RateLimits.Enabled {
+		delegate, err = NewRateLimitingRoundTripperWithOpts(
+			delegate, cfg.RateLimits.Limit, cfg.RateLimits.TransportOpts(),
+		)
+		if err != nil {
+			return nil, fmt.Errorf("create rate limiting round tripper: %w", err)
+		}
+	}
+
+	if cfg.Retries.Enabled {
+		opts := cfg.Retries.TransportOpts()
+		opts.LoggerProvider = providers.Logger
+		opts.BackoffPolicy = cfg.Retries.GetPolicy()
+		delegate, err = NewRetryableRoundTripperWithOpts(delegate, opts)
```

Comment:
A small suggestion – it might be better to apply the `RetryableRoundTripper` at the end of the chain to prevent other round trippers from being called during retries.